Source/WebCore/page/Frame.cpp

```diff
@@ -39,11 +39,64 @@
 namespace WebCore {
 
 #if ASSERT_ENABLED
-static HashMap<FrameIdentifier, WeakRef<Frame>>& allFrames()
-{
-    static NeverDestroyed<HashMap<FrameIdentifier, WeakRef<Frame>>> map;
-    return map;
-}
+class FrameLifetimeVerifier {
+public:
+    static FrameLifetimeVerifier& singleton()
+    {
+        static NeverDestroyed<FrameLifetimeVerifier> instance;
+        return instance.get();
+    }
+
+    void frameCreated(Frame& frame)
+    {
+        auto& pair = m_map.ensure(frame.frameID(), [] {
+            return std::pair<WeakPtr<LocalFrame>, WeakPtr<RemoteFrame>> { };
+        }).iterator->value;
+
+        switch (frame.frameType()) {
+        case Frame::FrameType::Local:
+            ASSERT_WITH_MESSAGE(!pair.first, "There should never be two LocalFrames with the same ID in the same process");
+            pair.first = downcast<LocalFrame>(frame);
+            break;
+        case Frame::FrameType::Remote:
+            ASSERT_WITH_MESSAGE(!pair.second, "There should never be two RemoteFrames with the same ID in the same process");
+            pair.second = downcast<RemoteFrame>(frame);
+            break;
+        }
+    }
+
+    void frameDestroyed(Frame& frame)
+    {
+        auto it = m_map.find(frame.frameID());
+        ASSERT(it != m_map.end());
+        auto& pair = it->value;
+        switch (frame.frameType()) {
+        case Frame::FrameType::Local:
+            ASSERT(pair.first == &frame);
+            if (pair.second)
+                pair.first = nullptr;
+            else
+                m_map.remove(it);
+            break;
+        case Frame::FrameType::Remote:
+            ASSERT(pair.second == &frame);
+            if (pair.first)
+                pair.second = nullptr;
+            else
+                m_map.remove(it);
+        }
+    }
+
+    bool isRootFrameIdentifier(FrameIdentifier identifier)
+    {
+        auto it = m_map.find(identifier);
+        if (it == m_map.end())
```

Comment:
Should we assert `it != m_map.end()`? Would web process check a frame ID that does not exist?

Reply:
It could check a FrameIdentifier from a Frame that has since been completely destroyed.  Returning false causes an assertion to fail for all current uses of isRootFrameIdentifier, but I could see why stronger assertions could be useful.